psl-core/src/main/java/org/linqs/psl/util/VizDataCollection.java

```diff
@@ -0,0 +1,215 @@
+import org.linqs.psl.model.rule.AbstractRule;
+import org.linqs.psl.model.term.Constant;
+import org.linqs.psl.model.term.Variable;
+
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.FilterOutputStream;
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.zip.GZIPOutputStream;
+
+public class VizDataCollection {
+    private static final Logger log = LoggerFactory.getLogger(VizDataCollection.class);
+
+    private static Runtime runtime = null;
+    private static VisualizationData vizData = null;
+
+    private static String outputPath = null;
+
+    static {
+        init();
+    }
+
+    private VizDataCollection() {}
+
+    private static synchronized void init() {
+        if (runtime != null) {
+            return;
+        }
+        vizData = new VisualizationData();
+        runtime = Runtime.getRuntime();
+        runtime.addShutdownHook(new ShutdownHook());
+    }
+
+    public static void outputJSON() throws IOException {
+        FilterOutputStream stream = System.out;
+
+        if (outputPath != null) {
+            try {
+                stream = new GZIPOutputStream(new PrintStream(outputPath));
+            } catch (IOException ex) {
+                throw new RuntimeException(ex);
+            }
+        }
+
+        writeToStream(stream);
+
+        if (outputPath != null) {
+            stream.close();
+        }
+    }
+
+    /**
+     * Write to stream with JSON formatting.
+     */
+    private static void writeToStream(FilterOutputStream stream) throws IOException {
+        // JSON format reference: https://www.json.org/json-en.html.
+        stream.write("{ \"truthMap\" :".getBytes());
+
+        // Write each map as a JSON object, each JSON object is comma delimited.
+        writeMap(stream, vizData.truthMap, "truthMap");
+        stream.write(", \"rules\" :".getBytes());
+        writeMap(stream, vizData.rules, "rules");
+        stream.write(", \"groundRules\" :".getBytes());
+        writeMap(stream, vizData.groundRules, "groundRules");
+        stream.write(", \"groundAtoms\" :".getBytes());
+        writeMap(stream, vizData.groundAtoms, "groundAtoms");
+
+        stream.write('}');
+    }
+
+    /**
+     * Write map to stream with JSON formatting.
+     */
+    @SuppressWarnings("unchecked")
+    private static void writeMap(FilterOutputStream stream, Object map, String z) throws IOException {
+        stream.write('{');
+
+        Map<String, Object> stringObjMap = (Map<String, Object>) map;
+        Iterator<Map.Entry<String, Object>> iterator = stringObjMap.entrySet().iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<String, Object> entry = iterator.next();
+            stream.write((" \"" + entry.getKey() + "\" :").getBytes());
```

Comment:
Avoid the additional whitespace inside the JSON.

Reply:
I thought JSON keys require a whitespace before and after the key followed by a colon, if we remove these whitespaces would the JSON file still be able to be parsed by the JSON parser?

Reply:
Instead of asking me, just test it out.